client/lib/domains/resolve-domain-status.js

```diff
@@ -0,0 +1,209 @@
+/**
+ * External dependencies
+ */
+import { translate } from 'i18n-calypso';
+import moment from 'moment';
+
+/**
+ * Internal dependencies
+ */
+import { transferStatus, type as domainTypes } from './constants';
+import { isExpiringSoon } from 'lib/domains/utils/is-expiring-soon';
+import { isRecentlyRegistered } from 'lib/domains/utils/is-recently-registered';
+import { hasPendingGSuiteUsers } from 'lib/gsuite';
+import { shouldRenderExpiringCreditCard } from 'lib/purchases';
+
+export function resolveDomainStatus(
+	domain,
+	purchase = null,
+	isJetpackSite = null,
+	isSiteAutomatedTransfer = null
+) {
+	switch ( domain.type ) {
+		case domainTypes.MAPPED:
+			if ( isExpiringSoon( domain, 30 ) ) {
+				const expiresMessage = translate( 'Expires in %(days)s', {
+					args: { days: moment.utc( domain.expiry ).fromNow( true ) },
+				} );
+
+				if ( isExpiringSoon( domain, 5 ) ) {
+					return {
+						statusText: expiresMessage,
+						statusClass: 'status-error',
+						icon: 'info',
+						listStatusText: expiresMessage,
+						listStatusClass: 'alert',
+					};
+				}
+
+				return {
+					statusText: expiresMessage,
+					statusClass: 'status-warning',
+					icon: 'info',
+					listStatusText: expiresMessage,
+					listStatusClass: 'warning',
+				};
+			}
+
+			if ( ( ! isJetpackSite || isSiteAutomatedTransfer ) && ! domain.pointsToWpcom ) {
+				return {
+					statusText: translate( 'Action required' ),
+					statusClass: 'status-error',
+					icon: 'info',
+				};
+			}
+
+			if ( hasPendingGSuiteUsers( domain ) ) {
+				return {
+					statusText: translate( 'Action required' ),
+					statusClass: 'status-warning',
+					icon: 'info',
+				};
+			}
+
+			return {
+				statusText: translate( 'Active' ),
+				statusClass: 'status-success',
+				icon: 'check_circle',
+			};
+
+		case domainTypes.REGISTERED:
+			if ( domain.pendingTransfer ) {
+				return {
+					statusText: translate( 'Outbound transfer initiated' ),
+					statusClass: 'status-error',
+					icon: 'cached',
+				};
+			}
+
+			if ( purchase && shouldRenderExpiringCreditCard( purchase ) ) {
+				return {
+					statusText: translate( 'Action required' ),
+					statusClass: 'status-error',
+					icon: 'info',
+				};
+			}
+
+			if ( domain.isPendingIcannVerification && domain.currentUserCanManage ) {
+				return {
+					statusText: translate( 'Action required' ),
+					statusClass: 'status-error',
+					icon: 'info',
+				};
+			}
+
+			if ( domain.expired ) {
+				return {
+					statusText: translate( 'Action required' ),
+					statusClass: 'status-error',
+					icon: 'info',
+					listStatusText: translate( 'Expired %(timeSinceExpiry)s', {
```

Comment:
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **29** times:
`translate( 'Expired %(timeSinceExpiry)s', { context: 'timeSinceExpiry is of the form "[number] [time-period] ago" e.g. "3 days ago"'} )` *ES Score: 12*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/domain-status-in-all-domains-list)